saleor/payment/models.py

```diff
@@ -0,0 +1,176 @@
+        validators=[MinValueValidator(1), MaxValueValidator(12)],
+        null=True, blank=True)
+    # exp year should be in 4 digits format
+    cc_exp_year = models.PositiveIntegerField(
+        validators=[MinValueValidator(1000)], null=True, blank=True)
+
+    def __repr__(self):
+        return 'Payment(gateway=%s, is_active=%s, created=%s, charge_status=%s)' % (
+            self.gateway, self.is_active, self.created, self.charge_status)
+
+    def get_last_transaction(self):
+        return max(self.transactions.all(), default=None, key=attrgetter('pk'))
+
+    def get_total(self):
+        return Money(self.total, self.currency or settings.DEFAULT_CURRENCY)
+
+    def get_captured_amount(self):
+        return Money(
+            self.captured_amount, self.currency or settings.DEFAULT_CURRENCY)
+
+    def authorize(self, payment_token):
+        from . import utils
+        return utils.gateway_authorize(
+            payment=self, payment_token=payment_token)
+
+    def void(self):
+        from . import utils
+        return utils.gateway_void(payment=self)
+
+    def capture(self, amount=None):
+        from . import utils
+        if amount is None:
+            # If no amount is specified, capture the maximum possible
+            amount = self.total - self.captured_amount
+        return utils.gateway_capture(payment=self, amount=amount)
+
+    def refund(self, amount=None):
+        from . import utils
+        if amount is None:
+            # If no amount is specified, refund the maximum possible
+            amount = self.captured_amount
+        return utils.gateway_refund(payment=self, amount=amount)
+
+    def can_authorize(self):
+        return (
+            self.is_active and self.charge_status == ChargeStatus.NOT_CHARGED)
+
+    def can_capture(self):
+        # FIXME should also have an auth transaction
+        not_charged = self.charge_status == ChargeStatus.NOT_CHARGED
+        not_fully_charged = (
+            self.charge_status == ChargeStatus.CHARGED
+            and self.get_total() > self.get_captured_amount())
+        return self.is_active and not_charged or not_fully_charged
+
+    def can_void(self):
+        return (
+            self.is_active and self.charge_status == ChargeStatus.NOT_CHARGED)
+
+    def can_refund(self):
+        return (
+            self.is_active and self.charge_status == ChargeStatus.CHARGED
+            and self.gateway != CustomPaymentChoices.MANUAL)
+
+
+class Transaction(models.Model):
+    """Transaction represent attempts to transfer money between your store
+    and your customers, with a chosen payment method.
+    """
+    created = models.DateTimeField(auto_now_add=True, editable=False)
+    payment = models.ForeignKey(
+        Payment, related_name='transactions', on_delete=models.PROTECT)
+    token = models.CharField(max_length=128, blank=True, default='')
+    kind = models.CharField(max_length=10, choices=TransactionKind.CHOICES)
```

Comment:
I think we should index this field or add a flag instead of `CharField`, as we use it to filter transactions when we are refunding.